### DISCUSSION TITLE
Lack of Thrust in new turbine creation

### DISCUSSION THREAD

**Discussion by Jlewis17** [NatLabRockies/floris#854] [Q&A] [answered]
Hi,

This may be a silly question but is there any way to implement a turbine into the model when I only have the power curve and turbine specifications? So I don't have a thrust coefficient value or dataframe as seen in the examples.

Many Thanks.

**Comment by misi9170** [comment]
Hi @Jlewis17 , 

Floris requires the user to specify a thrust coefficient, so you'll need to approximate the thrust coefficient for your turbine. Here are a couple of options that might work for you to do this:
1. Simply copy the thrust coefficient (Ct) curve from one of the turbines in the [turbine_library](https://github.com/NREL/floris/tree/main/floris/turbine_library). Because the thrust coefficient is nondimensional, you can select any of them you like, but perhaps it makes sense to copy the one from the turbine that is the closest match to your specifications (e.g. rated wind speed, rated power, or rotor diameter).
2. Use actuator disk theory (e.g https://www.e-education.psu.edu/aersp583/node/470, or other sources) to convert your power curve to a power coefficient; then to an axial induction factor; and then to an equivalent thrust coefficient. 
3. If you have a higher fidelity model, you may be able to determine the thrust coefficient empirically.

Of these, and assuming you don't have a high-fidelity turbine model, I would recommend the first option. The issue with the second option is that it assumes that the turbine is perfectly efficient, so will result in a lower thrust coefficient than what is practically observed. To determine which of the reference turbines to base your thrust curve off, consider looking at example 18, which plots the thrust curves of each of the reference turbines (as well as other behavior), producing:

![thrust_curves](https://github.com/NREL/floris/assets/39596329/3c097b76-dc46-447b-8cb9-9385f00d65ac)


As you can see, the thrust curves for the NREL 5MW, IEA 10MW, and IEA 15MW are not drastically different, even though their rated powers differ greatly. I would avoid using the x_20MW model---we don't have good data for that, and it is being removed in the upcoming release of FLORIS v4 (the thrust curves for the other models will also be changing slightly at that point)---although even the x_20MW has a similar Ct curve.

**Reply by Jlewis17** [reply]
Hi @misi9170,

Thanks for the information! I was looking at the following equations taken from Frohboese & Schmuck (2010) 

(1) Between cut-in (𝑈𝑖𝑛) and rated wind speed (𝑈𝑅) this method may be followed
𝐶𝑇 =3.5[𝑚𝑠](2𝑈+3.5[𝑚𝑠])/𝑈^2 ≈7[𝑚𝑠]/𝑈

(2) After rated wind speed, when the pitch control is active, the power is assumed to be kept constant, and
thus the thrust coefficient is expressed as 𝐶𝑇 = 3.5 [𝑚𝑠]𝑈𝑅 (2𝑈 + 3.5 [𝑚𝑠]) ∙1𝑈^3 ≈ 7 [𝑚𝑠] ∙𝑈R^2/𝑈^3

(3) The thrust coefficient is assumed not to exceed 1, therefore in the low wind speed regime where the
formula of Frohboese & Schmuck (2010) overestimates the thrust coefficient, the value is capped at 1.

Let me know if you think that would work?

**Reply by misi9170** [reply]
Hi @Jlewis17 ,

I tried to implement what you suggested, but I don't think I understood the equation for above-rated operation correctly (and I could not find it in [Frohboese & Schmuck](https://www.semanticscholar.org/paper/THRUST-COEFFICIENTS-USED-FOR-ESTIMATION-OF-WAKE-FOR-Frohboese-Hassan/6f1fdaec4f05daeb839c52e11109e403cf645e0a), but perhaps this is not the paper you were referring to---could you share a link to the correct paper?

The code I implemented is the following (for the NREL 5MW, which has rotor diameter 126m):
```python
        ws_r = 11.4 # Rated wind speed
        ws_c = 3.0 # Approximate cut in wind speed

        ws_arr = np.array(fi.floris.farm.turbine_map[0].power_thrust_table["wind_speed"])

        Cts = np.where(ws_arr < ws_c, 0.0, 7.0/ws_arr)
        Cts = np.where(ws_arr > ws_r, 7.0*ws_arr*(126/2)**2/ws_arr**3, Cts) # This doesn't seem right
        Cts = np.minimum(Cts, 1.0)
        axarr_cp_ct[1].plot(ws_arr,Cts,label='Heuristic', linestyle="dashed", color="black")
```

This produced the following curve (black dashed line):

![thrust_curves_2](https://github.com/NREL/floris/assets/39596329/cd3d72d3-4945-4116-91e6-d8298a41e82a)

Obviously I have something wrong above rated. But even below rated, this doesn't seem to be a great approximation of the actual thrust curve of the NREL 5MW model.

**Reply by Jlewis17** [reply]
Hi @misi9170,

Thanks for looking into this for me. The paper I got it from was here: https://core.ac.uk/download/pdf/96779884.pdf
"Arany, L., Bhattacharya, S. , Macdonald, J., & Hogan, J. (2017). Design of
monopiles for offshore wind turbines in 10 steps"

I think they got the original equation from equation 3 in the paper you have linked.

However, looking at your results, it doesn't look like an appropriate approximation.

**Reply by misi9170** [reply, marked as answer]
Hi @Jlewis17 , Ok, I see---from eqs (11) and (12) from the Arany et al. paper, implemented as 
```python
        ws_r = 11.4
        ws_cin = 3.0
        ws_cout = 25.0

        ws_arr = np.array(fi.floris.farm.turbine_map[0].power_thrust_table["wind_speed"])

        Cts = np.where(ws_arr < ws_cin, 0.0, 7.0/ws_arr)
        Cts = np.where(ws_arr > ws_r, 7.0*ws_r**2/ws_arr**3, Cts)
        Cts = np.where(ws_arr > ws_cout, 0.0, Cts)
        Cts = np.minimum(Cts, 1.0) # Clip to 1.0
        axarr_cp_ct[1].plot(ws_arr,Cts,label='Heuristic', linestyle="dashed", color="black")
```
I now get
![thrust_curves_3](https://github.com/NREL/floris/assets/39596329/2f1e5fa7-9e9f-4666-aa99-15602b4094d8)

This doesn't look too bad, but there is still a bit of a mismatch below rated and also a fairly large mismatch around the rated speed. Given that, I would probably stick with just taking the power curve from one of the existing models, but that's just my opinion :)

**Reply by Jlewis17** [reply]
Hi @misi9170,

Interesting, thanks for looking into this for me! You've been extremely helpful.
I think I will try and match the power curves as you suggested, however I may find some of my examples are too distant in which case I may have to use those equations.

Many Thanks!
